### DISCUSSION TITLE
Composite action inputs inside a shell script

### DISCUSSION THREAD

**Discussion by 38ri581oq480** [community/community#27088] [Actions]
<p>Hello!<br>
How can I access composite action inputs inside a shell script?</p>
<pre><code class="lang-auto">${{ inputs.input_name }}
</code></pre>
<p>and</p>
<pre><code class="lang-auto">$INPUT_INPUT_NAME
</code></pre>
<p><em><strong>not working inside a shell script!</strong></em></p>
<p>For example:<br>
<strong>action.yml</strong></p>
<pre><code class="lang-yaml">inputs:
  test_input:
    name: "Test input"
runs:
  using: 'composite'
  steps:
    - run: ${{ github.action_path }}/entrypoint.sh
      shell: bash
</code></pre>
<p><strong>entrypoint.sh</strong></p>
<pre><code class="lang-shell">echo ${{ inputs.test_input }} 
echo $INPUT_TEST_INPUT
</code></pre>

**Comment by airtower-luna** [comment, marked as answer]
<p>I’ve had that issue, too, unfortunately the bug report is still unresolved:</p>
<aside class="onebox githubissue">
  <header class="source">

      <a href="https://github.com/actions/runner/issues/665" target="_blank" rel="noopener">github.com/actions/runner</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/runner/issues/665" target="_blank" rel="noopener">INPUT_* environment variables are missing in composite actions</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-08-17" data-time="20:54:00" data-timezone="UTC">08:54PM - 17 Aug 20 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/airtower-luna" target="_blank" rel="noopener">
          <img alt="airtower-luna" src="https://user-images.githubusercontent.com/5814919/181101633-f413e1be-335a-44cb-9413-dd90e536743f.png" class="onebox-avatar-inline" width="20" height="20">
          airtower-luna
        </a>
      </div>
    </div>

    <div class="labels">
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">documentation</span>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">[According to the documentation](https://docs.github.com/en/actions/creating-act<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">ions/metadata-syntax-for-github-actions#inputs) actions should receive their input values as environment variables prefixed with `INPUT_` in addition to the `inputs` context. These environment variables are missing for composite actions.

**To Reproduce**
Steps to reproduce the behavior:
1. Create a composite action with one or more inputs, and a step that outputs the associated `INPUT` variables:
```yaml
inputs:
  foo:
    description: 'Test input'
    required: yes
runs:
  using: composite
  steps:
    - run: |
        echo "FOO: ${INPUT_FOO}"
      shell: bash
```
2. Use the action in a workflow:
```yaml
steps:
  - uses: ./example-action
    with:
      foo: 'bar'
```
3. Observe the lack of value after the `FOO:` prefix.

I have an [example action](https://github.com/airtower-luna/hello-github-actions/blob/master/composite-action/action.yaml) and [workflow](https://github.com/airtower-luna/hello-github-actions/blob/master/.github/workflows/main.yml) (the `uses: ./composite-action` step) that demonstrate the issue and the obvious workaround. There's a [workflow run](https://github.com/airtower-luna/hello-github-actions/runs/995382506) as well (check the "Run /./composite-action" step).

**Expected behavior**
Steps should be able to read inputs via environment variables, `INPUT_FOO` in the example above, `INPUT_NUM1` (to 4) in my reproducer.

## Runner Version and Platform
Version of your runner?
* 2.272.0

OS of the machine running the runner? OSX/Windows/Linux/...
* Ubuntu 20.04.1 LTS (Github hosted runner VM)

## Job Log Output
From the workflow run linked above:
```
Run ./composite-action
  with:
    num1: 1
    num2: 4
    num3: 0
    num4: 0
Show INPUT vars
  INPUT_NUM1 = 
  INPUT_NUM2 = 
  INPUT_NUM3 = 
  INPUT_NUM4 = 
Add numbers
```

## Credit
I became aware of the issue thanks to a [post on the Github community forum](https://github.community/t/input-variable-name-is-not-available-in-composite-run-steps/127611) and tried to reproduce the issue out of curiosity.</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Use <code>env</code> as a workaround, e.g.:</p>
<pre><code class="lang-yaml">    - run: ${{ github.action_path }}/entrypoint.sh
      shell: bash
      env:
        INPUT_TEST_INPUT: ${{ inputs.test_input }}
</code></pre>

**Comment by 38ri581oq480** [comment]
<p>thanks for workaround! :star_struck:<br>
Is it safe to pass secrets in <code>env</code> from <code>inputs</code>?<br>
For example:</p>
<pre><code class="lang-yaml">- run: ${{ github.action_path }}/entrypoint.sh
  shell: bash
  env:
    SUPER_SECRET: ${{ inputs.super_secret }}
</code></pre>

**Reply by lazarillo** [reply]
This seems problematic on two fronts:
1. You don't want to be passing secrets to `inputs`, but instead to `secrets`.  Something [like this](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions?tool=cli#using-secrets-in-a-workflow).
2. As the link shows, you have now made `$SUPER_SECRET` available as an environment variable.  So you'll have to make sure not to perform any output that shows its value.

The "trick" with using secrets in a bash environment is that you have to think of the cases where the secret _might_ get leaked.  For instance, if you did something like this:

```
python -m my_module param1 "$SUPER_SECRET"
```

Then if your program crashes or has an error, it is _quite possible_ that the value in `$SUPER_SECRET` would be output as part of the stack trace (depending upon how it's being used in the module).  Information can "leak" out when errors occur.  And if the secret variable is no longer _considered_ as secret (it's now just a normal environment variable), GH Actions cannot create any mechanisms to ensure that it isn't leaked out during output in a traceback or whatever.

**Reply by airtower-luna** [reply]
> You don't want to be passing secrets to `inputs`, but instead to `secrets`. Something [like this](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions?tool=cli#using-secrets-in-a-workflow).

Passing secrets to `inputs` is exactly what the example in the linked documentation does. An action (composite or otherwise) doesn't have direct access to `secrets`, it'll only receive those passed via inputs or environment.

> And if the secret variable is no longer considered as secret (it's now just a normal environment variable), GH Actions cannot create any mechanisms to ensure that it isn't leaked out during output in a traceback or whatever.

Secrets get masked in workflow logs no matter *how* they secret got into there. There's no "it's no longer considered a secret" if you set an environment variable from a secret value.

What could be a problem is if the secret got obfuscated in some way (e.g. base64 encoded) and *then* logged, and slips through the masking process that way. Or if the secret got written to a file and that file uploaded as an artifact. Or if you call a malicious action that uploads a secret you gave it to somewhere. And so on. :wink:

**Comment by airtower-luna** [comment]
<p>The <code>SUPER_SECRET</code> environment variable will be available to everything called from the command(s) defined in the <code>run</code> parameter. In this case that’s your script and anything it calls (unless the script overrides the environment variable). So it depends on your script. :slightly_smiling_face:</p>
